File: cpp/MP4Atom.cpp
//
//  MP4Atom.cpp
//  QTParse
//
//  Created by Alex Restrepo on 11/17/16.
//

#include "MP4Atom.h"

#include <cassert>
#include <algorithm>

MP4Atom::~MP4Atom() {

}

MP4Atom::MP4Atom()
: _offsetInFile(-1)
, _type(0)
, _hasExtendedSize(false)
, _size(0)
, _parent(nullptr)
, _children()
, _payload(nullptr)
, _payloadSize(0)
{
    
}

MP4Atom::MP4Atom(uint32_t type)
: MP4Atom() {
    setType(type);
}

MP4Atom MP4Atom::atomWithComponents(off_t offsetInFile, bool hasExtendedSize, uint64_t size, uint32_t type, MP4Atom *parent) {
    MP4Atom atom(type);
    atom._offsetInFile = offsetInFile;
    atom._hasExtendedSize = hasExtendedSize;
    atom._size = size;
    atom._parent = parent;
    
    return atom;
}

MP4Atom MP4Atom::atomWithPayload(uint32_t type, void *payload, uint64_t payloadSize) {
    MP4Atom atom(type);
    atom._payloadSize = payloadSize;
    atom._size = atom.headerSize() + payloadSize;
    atom._payload = payload;
    
    return atom;
}

uint8_t MP4Atom::headerSize() {
    return _hasExtendedSize ? 16 : 8;
}

off_t MP4Atom::endOffsetInFile() {
    return _offsetInFile + _size;
}

std::string StringFromType(uint32_t type) {
    unsigned char strType[5];
    strType[0] = (type >> 24) & 0xff;
    strType[1] = (type >> 16) & 0xff;
    strType[2] = (type >> 8) & 0xff;
    strType[3] = type & 0xff;
    strType[4] = 0;
    
    return std::string(reinterpret_cast<const char*>(strType));
}

void MP4Atom::setType(uint32_t type) {
    _type = type;
    _typeAsString = StringFromType(type);
}

uint64_t MP4Atom::calculatedSize() {
    if (!_children.size()) {
        return _size;
    }
    
    uint64_t size = 0;
    for (MP4Atom &atom : _children) {
        size += atom.calculatedSize();
    }
    
    return size + headerSize();
}

std::vector<MP4Atom *> MP4Atom::childrenWithType(uint32_t type) {
    std::vector<MP4Atom *> result;
    for (MP4Atom &atom : _children) {
        if (atom._type == type) {
            result.push_back(&atom);
        }
    }
    
    return result;
}

void MP4Atom::printDescription() {
    printf("['%s'][%lld + %llu%s(%lld) = %lld]", _typeAsString.c_str(), _offsetInFile, _size, _hasExtendedSize ? "[ext]" : "", calculatedSize(), endOffsetInFile());
}

//void MP4Atom::printStructure() {
//    printStructureWithPadding("");
//}

void _replaceAll(std::string& str, const std::string& from, const std::string& to) {
    if(from.empty())
        return;
    size_t start_pos = 0;
    while((start_pos = str.find(from, start_pos)) != std::string::npos) {
        str.replace(start_pos, from.length(), to);
        start_pos += to.length(); // In case 'to' contains 'from', like replacing 'x' with 'yx'
    }
}

void MP4Atom::printStructureWithPadding(const char *padding) {
    printf("%s", padding);
    printDescription();
    printf("\n");
    
    for (int i = 0; i < _children.size(); i++) {
        MP4Atom atom = _children[i];
        std::string nextPadding(padding);
        _replaceAll(nextPadding, "├", "│");
        _replaceAll(nextPadding, "└", " ");
        _replaceAll(nextPadding, "─", " ");
        
        if (i < _children.size() - 1) {
            nextPadding = nextPadding.append("  ├─");
        } else {
            nextPadding = nextPadding.append("  └─");
        }
        
        atom.printStructureWithPadding(nextPadding.c_str());
    }
}

void MP4Atom::appendAtom(MP4Atom atom) {
    if (atom.size() < 8) {
        return;
    }
    
    atom._parent = this;
    _children.push_back(atom);
}

void MP4Atom::loadPayload(FILE *source) {
    // does nothing by default.
}

size_t MP4Atom::appendPayloadToFile(FILE *file) {
    size_t written = 0;
    if (_payload) {
        written = fwrite(_payload, 1, _payloadSize, file);
    }
    return written;
}

File: objc/MP4Atom.m
//
//  MP4Atom.m
//  MP4Parser
//
//  Created by Alex Restrepo on 2/11/16.
//

#import "MP4Atom.h"
#import "NSString+AtomType.h"

static NSMutableArray *registeredAtoms = nil;

@implementation NSFileHandle(data)
- (NSData *)he_dataWithRange:(NSRange)range {
    [self seekToFileOffset:range.location];
    NSData *currentData = [self readDataOfLength:range.length];
    return currentData;
}

- (void)he_getBytes:(void *)buffer range:(NSRange)range {
    NSData *currentData = [self he_dataWithRange:range];
    [currentData getBytes:buffer range:NSMakeRange(0, range.length)];
}

@end

@implementation MP4Atom

// http://www.cocoawithlove.com/2009/07/simple-extensible-http-server-in-cocoa.html
+ (void)registerAtom:(Class)atomClass {
    if (registeredAtoms == nil) {
        registeredAtoms = [[NSMutableArray alloc] init];
    }
    [registeredAtoms addObject:atomClass];
}

+ (MP4Atom *)atomForType:(uint32_t)type {
    MP4Atom *atom = nil;
    
    for(Class atomClass in registeredAtoms) {
        if(type == [atomClass atomClassType]) {
            atom = [[atomClass alloc] initWithType:type];
            break;
        }
    }
    
    if (!atom) {
        atom = [[MP4Atom alloc] initWithType:type];
    }
    
    return atom;
}

+ (uint32_t)atomClassType {
    return 0;
}

- (instancetype)init {
    self = [self initWithType:[[self class] atomClassType]];
    if (self) {
        
    }
    return self;
}

- (instancetype)initWithType:(uint32_t)type
{
    self = [super init];
    if (self) {
        _offsetInFile = -1;
        _type = type;
        _headerSize = 8; //default
    }
    return self;
}

- (off_t)endOffsetInFile {
    return _offsetInFile + MAX(_size, _size);
}

- (uint64_t)dataSize {
    return _size - _headerSize;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"['%@'][%lld + %llu%@(%lld): %lld]",// [%p-%p]",
            [NSString stringWithAtomType:_type], _offsetInFile, _size, _hasExtendedSize ? @"[ext]" : @"", [self calculatedSize], [self endOffsetInFile] /*, self, _parent*/];
}

- (NSString *)descriptionWithPadding:(NSString *)padding {
    NSMutableString *description = [NSMutableString stringWithString:padding];
    [description appendFormat:@"%@\n",[self description]];

    [_children enumerateObjectsUsingBlock:^(MP4Atom * _Nonnull atom, NSUInteger idx, BOOL * _Nonnull stop) {
        NSString *nextPadding = padding;
        nextPadding = [nextPadding stringByReplacingOccurrencesOfString:@"├" withString:@"│"];
        nextPadding = [nextPadding stringByReplacingOccurrencesOfString:@"└" withString:@" "];
        nextPadding = [nextPadding stringByReplacingOccurrencesOfString:@"─" withString:@" "];
        
        if (idx < [_children count] - 1) {
            nextPadding = [nextPadding stringByAppendingString:@" ├─"];
            
        } else {
            nextPadding = [nextPadding stringByAppendingString:@" └─"];
        }
        
        [description appendString:[atom descriptionWithPadding:nextPadding]];
    }];
    
    return description;
}

- (void)processData:(NSFileHandle *)fileHandle {
    // ignore by default
}

- (NSArray *)atomsWithType:(uint32_t)type {
    NSMutableArray *result = [NSMutableArray new];
    for (MP4Atom *atom in _children) {
        if (atom.type == type) {
            [result addObject:atom];
        }
    }
    return [result copy];
}

- (void)appendAtom:(MP4Atom *)atom {
    if (!atom || atom.size < 8) {
        return;
    }
    
    // 1. append atom
    atom.parent = self;
    [self.children addObject:atom];


    /*
     We can mark atoms as "not saved" by giving them an
     offset of -1.
     Then when writing the new file, it will be written as a stream, so we just do a traversal of the structure and at that point
     we can calculate the new offsets...
    */
    atom.offsetInFile = -1;
}

- (void)reconcileOffsets {
    off_t startOffset = self.offsetInFile + self.headerSize;

    if (![self.children count]) {
        return;
    }
    
    for (MP4Atom *atom in self.children) {
        atom.offsetInFile = startOffset;
        [atom reconcileOffsets];
        
        startOffset += atom.size;
    }
    
    self.size = [[self.children  lastObject] endOffsetInFile] - self.offsetInFile;
}

- (uint64_t)calculatedSize {
    if (![self.children count]) {
        return self.size;
    }
    
    uint64_t size = 0;
    for (MP4Atom *atom in _children) {
        size += [atom calculatedSize];
    }
    
    return size + self.headerSize;
}

- (off_t)calculatedOffsetInFile {
    __block off_t offset = [_parent calculatedOffsetInFile];
    [_parent.children enumerateObjectsUsingBlock:^(MP4Atom * _Nonnull sibling, NSUInteger idx, BOOL * _Nonnull stop) {
        if (sibling == self) {
            *stop = YES;
        } else {
            offset += [sibling calculatedSize];
        }
    }];

    return offset;
}

- (NSInteger)appendContentsToStream:(NSOutputStream *)stream {
    // default does nothing, this will make the save method fail in an atom was modified and doesn't implement this method.
    return 0;
}

@end
